src/demikernel/dbuf.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+//==============================================================================
+// Imports
+//==============================================================================
+
+use ::runtime::{
+    fail::Fail,
+    memory::Buffer,
+};
+use ::std::{
+    fmt,
+    fmt::Debug,
+    ops::{
+        Deref,
+        DerefMut,
+    },
+    slice,
+    sync::Arc,
+};
+
+//==============================================================================
+// Structures
+//==============================================================================
+
+/// Data Buffer.
+#[derive(Clone)]
+pub struct DataBuffer {
+    /// Underlying data.
+    data: Option<Arc<[u8]>>,
+
+    /// Data offset.
+    offset: usize,
+
+    /// Data length
+    len: usize,
+}
+
+//==============================================================================
+// Associated Functions
+//==============================================================================
+
+/// Associated Functions for Data Buffers
+impl DataBuffer {
+    // Creates a data buffer with a given capacity.
+    pub fn new(capacity: usize) -> Result<Self, Fail> {
+        // Bad capacity.
+        if capacity == 0 {
+            return Err(Fail::new(libc::EINVAL, "zero-capacity buffer"));
+        }
+
+        // Create buffer.
+        Ok(Self {
+            data: unsafe { Some(Arc::new_zeroed_slice(capacity).assume_init()) },
+            offset: 0,
+            len: capacity,
+        })
+    }
+
+    /// Creates a data buffer from a raw pointer and a length.
+    pub fn from_raw_parts(data: *mut u8, len: usize) -> Result<Self, Fail> {
+        // Bad length.
+        if len == 0 {
+            return Err(Fail::new(libc::EINVAL, "zero-length buffer"));
+        }
+
```

Comment:
I wonder if we should attempt some sort of raw pointer verification here.  Sure, it's hard to tell if a pointer is valid (at least without attempting to dereference it), and zero (a.k.a. null) is a valid pointer value on some systems.  But we might want to check against zero anyways, as it most likely would be an error.  In the interest of catching bugs fast.

What do you think?

Reply:
I've introduced null-pointer checking.

Reply:
Great, thanks.